[Woo POS] [WIP] Show inline message for payment alerts

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Based on Josh's https://github.com/woocommerce/woocommerce-ios/pull/13040

This PR shows that we can split the alert details into 2 different presentation types to be shown as "alert" or "message". It maps to the corresponding type in `CardPresentPaymentsAlertPresenterAdaptor`. There are a few TODO comments but they could potentially be done separately since all changes are affecting POS only.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Prerequisite: the store is eligible for POS and the feature switch is enabled in Menu > Settings > Experimental Features > POS
* Launch app
* Go to Menu > Point of Sale
* Add some item(s) to cart, then tap `Collect payment` --> in the happy path, the connection alerts should be shown as a sheet while payment events are shown inline in the totals view (as a temporary Text view with an orange background)

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/1945542/d3bb599f-b8db-47d9-818e-6c958e091bba



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
WIP changes while battery is low.

jaclync committed:
Remove unused code and add TODO comments.

jaclync committed:
Make the temporary inline message UI more noticeable with TODO comments to remove them.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented on WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentEvent.swift:
```diff
@@ -1,8 +1,28 @@
+    case showPaymentMessage(_ message: CardPresentPaymentMessageType)
     case showReaderList(_ readerIDs: [String], selectionHandler: ((String) -> Void))
     case showOnboarding(_ onboardingViewModel: CardPresentPaymentsOnboardingViewModel)
 }
```
We should go ahead with this approach if it's easiest, to move towards M1/MVP... but I think I made a mistake in the initial design of this enum.

Really, the service shouldn't dictate or suggest how the messages are displayed, that should be up to the caller. Perhaps we collapse all the possible messages into this single enum... or we could just change some names:

```
enum CardPresentPaymentEvent {
    case idle
    case showConnectionMessage(_ message: CardPresentPaymentConnectionMessageType) // include the reader list here too.
    case showPaymentMessage(_ message: CardPresentPaymentMessageType)
    case showOnboarding(_ onboardingDetails: CardPresentPaymentsOnboardingDetailsType)
}
```

joshheald commented on WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentsAlertPresenterAdaptor.swift:
```diff
@@ -14,7 +14,64 @@ final class CardPresentPaymentsAlertPresenterAdaptor: CardPresentPaymentAlertsPr
-        paymentAlertSubject.send(.showAlert(viewModel))
+        switch viewModel {
+            case .scanningForReaders(let endSearch):
+                // TODO: hide it in collectPayment flow
```
I don't think we need to hide these in the collect payment flow, as it should only start if the reader is already connected, so it won't get this scanning message.

Or is this in the case where the reader isn't connected when they get to the checkout, and the merchant taps a button to connect? I was expecting that we'd still show those modally, at least in M1, but I guess we don't have to. Surely we still want to say (inline) "scanning, please turn your reader on", so that the merchant knows it's doing something and that they need to switch it on...

joshheald commented on WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentsAlertPresenterAdaptor.swift:
```diff
@@ -14,7 +14,64 @@ final class CardPresentPaymentsAlertPresenterAdaptor: CardPresentPaymentAlertsPr
+                paymentAlertSubject.send(.showAlert(.connectingFailed(viewModel: .init())))
+            case .preparingForPayment(let cancelPayment):
+                paymentAlertSubject.send(.showPaymentMessage(.preparingForPayment))
+            // TODO: support this case
```
Or... delete it everywhere. The existing app doesn't actually use it.

joshheald commented on WooCommerce/Classes/POS/Presentation/CardPresentPaymentAlert.swift:
```diff
@@ -23,21 +23,6 @@ struct CardPresentPaymentAlert: View {
-            Text("Success")
-        case .error(let alertViewModel):
-            Text("Error")
         }
```
_Noodling: Being able to do this is the obvious benefit of the separate `CardPresentPaymentEvent` enum cases for connection/payment. However, we might be able get a similar benefit with a single flat enum for all payment events if the associated `CardPresentPaymentAlertType` conformed to a protocol for connection or payment alerts..._

_Not sure, it might be a little complicated._

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -33,14 +54,11 @@ struct PointOfSaleDashboardView: View {
-                    .showReaderList,
-                    .showOnboarding:
-                Text(viewModel.cardPresentPaymentEvent.temporaryEventDescription)
+            // Might be the only way unless we make the type conform to `Identifiable`
```
Making it identifiable seems reasonable. It's one of those where just adding a UUID is fine, I think.

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -136,14 +138,32 @@ private extension PointOfSaleDashboardViewModel {
-                    .showOnboarding:
+            case .showAlert:
                 return true
+            case .showReaderList,
```
I think we should flatten `showReaderList` into `showAlert`, because we can do that now that we have control over the view models emitted, and it's just another part of the connection flow.

joshheald reviewed (approved):
Nice! I'll merge this when I've done all the others, but some suggestions which we can do now or later or not at all. Let me know what you think?

joshheald committed:
13036 use injected card payment configuration

joshheald committed:
13036 Remove unused onboardingPresenter

joshheald committed:
13039 Make generic order consistent

CollectOrderPaymentUseCase and PreflightController had differing orders for their alert providers

joshheald committed:
13039 make factory final

joshheald committed:
13039 reuse connection alert provider

joshheald committed:
13039 make alert provider property private

joshheald committed:
13040 Pass cancellation handler to view details

joshheald committed:
Merge branch 'trunk' into issue/12868-show-payment-alert-inline

# Conflicts:
#	WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentEvent.swift
#	WooCommerce/Classes/POS/Presentation/CardPresentPaymentAlert.swift
#	WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift

joshheald committed:
Merge branch 'trunk' into issue/12868-pr-feedback-fixes

# Conflicts:
#	WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentEvent.swift
#	WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift

joshheald committed:
Merge branch 'issue/12868-show-payment-alert-inline' into issue/12868-pr-feedback-fixes

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift

joshheald committed:
Merge branch 'trunk' into issue/12868-show-payment-alert-inline

# Conflicts:
#	WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentEvent.swift
#	WooCommerce/Classes/POS/Presentation/CardPresentPaymentAlert.swift
#	WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift

joshheald committed:
[Woo POS] Reader flow specific alerts pr feedback fixes (#13054)
